Mathlib/Probability/CondJensen.lean

```diff
@@ -0,0 +1,577 @@
+
+/-- Conditional expectation commutes with affine functions -/
+theorem condExpL1_comp_affine {α 𝕜 E : Type*}
+    [NormedAddCommGroup E] [CompleteSpace E] [NormedSpace ℝ E]
+    [RCLike 𝕜] [Module 𝕜 E] [ContinuousSMul 𝕜 E]
+    {m mα : MeasurableSpace α} (hm : m ≤ mα) {μ : Measure α} [IsFiniteMeasure μ]
+    {f : α → E} (hf_int : Integrable f μ) (T : E →L[𝕜] 𝕜) (a : ℝ) :
+    (fun x ↦ re (T (μ[f | m] x)) + a) =ᵐ[μ] μ[fun y ↦ re (T (f y)) + a | m] := by
+  let g := @reCLM 𝕜 (by infer_instance)
+  let h := restrictScalars ℝ T
+  have reTf_int : Integrable ((re ∘ T) ∘ f) μ := integrable_comp (comp g h) hf_int
+  have hp : (fun x ↦ re (T (μ[f | m] x)) + a) =ᵐ[μ]
+      (μ[(re ∘ T) ∘ f | m] + μ[(fun y ↦ a) | m]) := by
+    filter_upwards [condExpL1_comp_continuousLinearMap hm hf_int (comp g h)] with b hb
+    simpa [condExp_const hm a] using hb
+  exact hp.trans (condExp_add reTf_int (integrable_const a) m).symm
+
+/-- Conditional Jensen for separable spaces -/
+lemma conditional_jensen_of_separableSpace {α X : Type*}
+    [NormedAddCommGroup X] [NormedSpace ℝ X] [CompleteSpace X] [SeparableSpace X]
+    {m mα : MeasurableSpace α} (hm : m ≤ mα) {μ : Measure α} [IsFiniteMeasure μ]
+    {φ : X → ℝ} (hφ_cvx : ConvexOn ℝ Set.univ φ) (hφ_cont : LowerSemicontinuous φ)
+    {f : α → X} (hf_int : Integrable f μ) (hφ_int : Integrable (φ ∘ f) μ) :
+    ∀ᵐ a ∂μ, φ (μ[f | m] a) ≤ μ[φ ∘ f | m] a := by
+  rcases hφ_cvx.iSup_affine_eq_of_separableSpace (𝕜 := ℝ) hφ_cont with ⟨L, c, hp⟩
+  have py : ∀ᵐ a ∂μ, ∀ i : ℕ, re ((L i) (μ[f | m] a)) + c i
+    = μ[re ∘ (L i) ∘ f + (fun (b : α) ↦ (c i)) | m] a := by
+    rw [ae_all_iff]; intro i; apply condExpL1_comp_affine hm hf_int (L i) (c i)
+  have pz : ∀ᵐ a ∂μ, ∀ i : ℕ, (re ∘ (L i) ∘ f + (fun (b : α) ↦ (c i))) a ≤ (φ ∘ f) a := by
+    rw [ae_all_iff]; intro i; filter_upwards with a
+    rw [Function.comp_apply, ← (hp (f a)).2, Pi.add_apply, Function.comp_apply, Function.comp_apply]
+    apply le_ciSup (hp (f a)).1 i
+  have pw : ∀ᵐ a ∂μ, ∀ i : ℕ, μ[(re ∘ (L i) ∘ f + (fun (b : α) ↦ (c i))) | m] a
+    ≤ μ[φ ∘ f | m] a := by
+    rw [ae_all_iff]; intro i; apply condExp_mono
+    · let g := @reCLM ℝ (by infer_instance)
+      have reLif_int : Integrable (fun (a : α) ↦ (re ∘ (L i)) (f a)) μ
+        := integrable_comp (comp g (L i)) hf_int
+      exact Integrable.add reLif_int (integrable_const (c i))
+    · exact hφ_int
+    · exact ae_all_iff.mp pz i
+  filter_upwards [py, pw] with a
+  intro hy hw
+  rw [← (hp (μ[f | m] a)).right]
+  apply ciSup_le
+  intro i
+  rw [hy i]
+  apply hw i
+
+lemma Measurable.codRestrict {Ω X : Type*} [MeasurableSpace Ω] [MeasurableSpace X] {f : Ω → X}
+    (hf : Measurable f) {s : Set X} (h : ∀ y, f y ∈ s) :
+    Measurable (codRestrict f s h) := hf.subtype_mk
+
+/-- Conditional Jensen's inequality. -/
+theorem conditional_jensen {α X : Type*}
+    [NormedAddCommGroup X] [NormedSpace ℝ X] [CompleteSpace X]
+    {m mα : MeasurableSpace α} (hm : m ≤ mα) {μ : Measure α} [IsFiniteMeasure μ]
```

Comment:
Can you please add a TODO in the file docstring to say that this should possibly be generalized to sigma-finite measures?